src/nnet3/online-nnet3-decodable.cc

```diff
@@ -0,0 +1,210 @@
+    feat_dim_(input_feats->Dim()),
+    num_pdfs_(nnet.GetNnet().OutputDim("output")),
+    begin_frame_(-1) {
+  KALDI_ASSERT(opts_.max_nnet_batch_size > 0);
+  log_priors_ = nnet_.Priors();
+  KALDI_ASSERT((log_priors_.Dim() == 0 || log_priors_.Dim() == trans_model_.NumPdfs()) &&
+               "Priors in neural network must match with transition model (if exist).");
+
+  ComputeSimpleNnetContext(nnet_.GetNnet(), &left_context_, &right_context_);
+  log_priors_.ApplyLog();
+}
+
+
+
+BaseFloat DecodableNnet3Online::LogLikelihood(int32 frame, int32 index) {
+  ComputeForFrame(frame);
+  int32 pdf_id = trans_model_.TransitionIdToPdf(index);
+  KALDI_ASSERT(frame >= begin_frame_ &&
+               frame < begin_frame_ + scaled_loglikes_.NumRows());
+  return scaled_loglikes_(frame - begin_frame_, pdf_id);
+}
+
+
+bool DecodableNnet3Online::IsLastFrame(int32 frame) const {
+  KALDI_ASSERT(false && "Method is not imlemented");
+  return false;
+}
+
+int32 DecodableNnet3Online::NumFramesReady() const {
+  int32 features_ready = features_->NumFramesReady();
+  if (features_ready == 0)
+    return 0;
+  bool input_finished = features_->IsLastFrame(features_ready - 1);
+  if (opts_.pad_input) {
+    // normal case... we'll pad with duplicates of first + last frame to get the
+    // required left and right context.
+    if (input_finished) return subsampling(features_ready);
+    else return std::max<int32>(0, subsampling(features_ready - right_context_));
+  } else {
+    return std::max<int32>(0, subsampling(features_ready - right_context_ - left_context_));
+  }
+}
+
+int32 DecodableNnet3Online::subsampling(int32 num_frames) const {
+  return (num_frames) / opts_.frame_subsampling_factor;
+}
+
+void DecodableNnet3Online::ComputeForFrame(int32 subsampled_frame) {
+  int32 features_ready = features_->NumFramesReady();
+  bool input_finished = features_->IsLastFrame(features_ready - 1);  
+  KALDI_ASSERT(subsampled_frame >= 0);
+  if (subsampled_frame >= begin_frame_ &&
+      subsampled_frame < begin_frame_ + scaled_loglikes_.NumRows())
+    return;
+  KALDI_ASSERT(subsampled_frame < NumFramesReady());
+
+  int32 subsample = opts_.frame_subsampling_factor;
+
+  int32 input_frame_begin;
+  if (opts_.pad_input)
+    input_frame_begin = subsampled_frame * subsample  - left_context_;
+  else
+    input_frame_begin = subsampled_frame * subsample;
+  int32 max_possible_input_frame_end = features_ready /* - ( features_ready - right_context_) % subsample */;
+  if (input_finished && opts_.pad_input)
+    max_possible_input_frame_end += right_context_;
+  int32 input_frame_end = std::min<int32>(max_possible_input_frame_end,
+                                          input_frame_begin +
+                                          left_context_ + right_context_ +
+                                          opts_.max_nnet_batch_size);
+  KALDI_ASSERT(input_frame_end > input_frame_begin);
+  Matrix<BaseFloat> features(input_frame_end - input_frame_begin,
```

Comment:
Regarding re-use of buffers, etc.-- the input features are very small
relative to the size of the computation itself, and I don't like to
optimize things until it's proven that they are significant.



On Mon, Apr 11, 2016 at 9:47 AM, Ilya Platonova <notifications@github.com>
wrote:

> In src/nnet3/online-nnet3-decodable.cc
> <https://github.com/kaldi-asr/kaldi/pull/666#discussion_r59238618>:
>
> > +  int32 subsample = opts_.frame_subsampling_factor;
> > +
> > +  int32 input_frame_begin;
> > +  if (opts_.pad_input)
> > +    input_frame_begin = subsampled_frame * subsample  - left_context_;
> > +  else
> > +    input_frame_begin = subsampled_frame * subsample;
> > +  int32 max_possible_input_frame_end = features_ready /* - ( features_ready - right_context_) % subsample */;
> > +  if (input_finished && opts_.pad_input)
> > +    max_possible_input_frame_end += right_context_;
> > +  int32 input_frame_end = std::min<int32>(max_possible_input_frame_end,
> > +                                          input_frame_begin +
> > +                                          left_context_ + right_context_ +
> > +                                          opts_.max_nnet_batch_size);
> > +  KALDI_ASSERT(input_frame_end > input_frame_begin);
> > +  Matrix<BaseFloat> features(input_frame_end - input_frame_begin,
>
> Also my general concern about decoders is that, they tend not to reuse
> input/output buffers. Maybe this is point for optimization here. @danpovey
> <https://github.com/danpovey> what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/666/files/75f8b027a2e6b6b679d26d0b4b667cea631ee6d1#r59238618>
>